### ISSUE TITLE
Granting to multiple roles in postgres

### ISSUE THREAD

**Issue by blx** [sqlfluff/sqlfluff#1365]
## Expected Behaviour

Parsing and linting correctly handle a `GRANT` statement to multiple roles/grantees in postgres dialect.

From [postgres docs](https://www.postgresql.org/docs/current/sql-grant.html):
```
GRANT { { SELECT | INSERT | UPDATE | DELETE | TRUNCATE | REFERENCES | TRIGGER }
    [, ...] | ALL [ PRIVILEGES ] }
    ON { [ TABLE ] table_name [, ...]
         | ALL TABLES IN SCHEMA schema_name [, ...] }
    TO role_specification [, ...] [ WITH GRANT OPTION ]
```

## Observed Behaviour

Parse error:

```
❯ echo 'GRANT SELECT ON abc TO xyz, mno;' | sqlfluff parse --dialect postgres -
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        unparsable:                                           !! Expected: 'StatementSegment'
[L:  1, P:  1]      |            raw:                                              'GRANT'
[L:  1, P:  6]      |            whitespace:                                       ' '
[L:  1, P:  7]      |            raw:                                              'SELECT'
[L:  1, P: 13]      |            whitespace:                                       ' '
[L:  1, P: 14]      |            raw:                                              'ON'
[L:  1, P: 16]      |            whitespace:                                       ' '
[L:  1, P: 17]      |            raw:                                              'abc'
[L:  1, P: 20]      |            whitespace:                                       ' '
[L:  1, P: 21]      |            raw:                                              'TO'
[L:  1, P: 23]      |            whitespace:                                       ' '
[L:  1, P: 24]      |            raw:                                              'xyz'
[L:  1, P: 27]      |            comma:                                            ','
[L:  1, P: 28]      |            whitespace:                                       ' '
[L:  1, P: 29]      |            raw:                                              'mno'
[L:  1, P: 32]      |    statement_terminator:                                     ';'
[L:  1, P: 33]      |    newline:                                                  '\n'

==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'GRANT SELECT ON abc
                       | TO xyz, mno'


❯ echo 'GRANT SELECT ON abc TO xyz, mno;' | sqlfluff lint --dialect postgres -
== [stdin] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'GRANT
                       | SELECT ON abc TO xyz, mno'
All Finished 📜 🎉!
```

## Dialect

Postgres

## Version
```
❯ sqlfluff --version
sqlfluff, version 0.6.4

❯ python3 --version
Python 3.9.6
```

## Configuration

No custom configuration

**Comment by WittierDinosaur** [comment]
I've had a go at fixing this, it should just be a simple case of wrapping this in a Delimited() in dialect_ansi.py:

```
            OneOf(
                Ref("ObjectReferenceSegment"),
                Ref("FunctionSegment"),
                "PUBLIC",
            ),
```

But for the life of me I can't get it to accept a delimiter. I think there may be a bug in Delimited(), but I can't work out where it might be. @tunetheweb can you confirm whether this strange behaviour happens for you?

Some test cases:

```
GRANT CONNECT, CREATE, TEMP, TEMPORARY ON DATABASE my_db TO my_user, my_other_user WITH GRANT OPTION;
GRANT SELECT ON abc TO xyz, xyzaaaaa;
GRANT EXECUTE ON ALL ROUTINES IN SCHEMA my_schema TO my_user, my_other_user;
```

**Comment by tunetheweb** [comment]
Yeah I thought this was a simple fix too! I think I get why.

This works:

```py
                Bracketed(
                    Delimited(
                        Ref("ObjectReferenceSegment"),
                    ),
                ),
```

But obviously don't want the brackets.

So I think, without the brackets, the `ObjectReferenceSegment` matches the first one, and it carries on assuming the second one is part of the next expression.

I think this needs to be pulled out into it's own `Ref()` with a `terminator = OneOf("WITH", ...)` clause.

**Comment by tunetheweb** [comment]
Or we implement a `greedy` option for `Delimited`?

**Comment by WittierDinosaur** [comment]
The issue there is that the terminator is optional. Shouldn't delimited by greedy by default? Anything else is surely unexpected behaviour. I too tried that thing with bracketed, and saw it worked. I've tried all kinds of combinations, including adding keywords after, and the behaviour is just bizarre.

For example, change it to this:

```
            Sequence(
                Delimited(
                    Ref("ObjectReferenceSegment"),
                    Ref("FunctionSegment"),
                    "PUBLIC",
                ),
                "FROM"
            ),
```

With this test case

```
GRANT SELECT ON abc TO xyz, xyzaaaaa FROM;
```

That fails.

**Comment by WittierDinosaur** [comment]
I've found the problem.

More investigation for root cause, but having two Delimited() in the same statement is causing some kind of conflict. In my example above, replace the delimited chunk further up with:

```
                Sequence(
                    # Delimited(
                        OneOf(_global_permissions, _permissions),
                    #     delimiter=Ref("CommaSegment"),
                    # ),
                    "ON",
                    _objects,
                ),
```
The lower down delimited will now work as expected. There's something class-level that's interfering.

**Comment by tunetheweb** [comment]
On good find. This works then by giving a `terminator` clause to that first one so it nows when to end it:

```py
                Sequence(
                    Delimited(
                        OneOf(_global_permissions, _permissions),
                        delimiter=Ref("CommaSegment"),
                        terminator="ON",
                    ),
                    "ON",
                    _objects,
                ),
```

**Comment by WittierDinosaur** [comment]
I've added a bug for this behaviour, and I'll fire off a PR for the GRANT stuff shortly, I'll include an explicit terminator as you've done above.
